src/k8s/pkg/component/dns.go

```diff
@@ -0,0 +1,136 @@
+package component
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/canonical/k8s/pkg/snap"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+)
+
+func EnableDNSComponent(s snap.Snap, clusterDomain, serviceIP string, upstreamNameservers []string) error {
+	manager, err := NewManager(s)
+	if err != nil {
+		return fmt.Errorf("failed to get component manager: %w", err)
+	}
+
+	upstreamNameserver := "/etc/resolv.conf"
+	if clusterDomain == "" {
+		clusterDomain = "cluster.local"
+	}
+
+	if len(upstreamNameservers) > 0 {
+		upstreamNameserver = strings.Join(upstreamNameservers, " ")
+	}
+
+	values := map[string]any{
+		"servers": []map[string]any{
+			{
+				"zones": []map[string]any{
+					{"zone": "."},
+				},
+				"port": 53,
+				"plugins": []map[string]any{
+					{"name": "errors"},
+					{"name": "health", "configBlock": "lameduck 5s"},
+					{"name": "ready"},
+					{
+						"name":        "kubernetes",
+						"parameters":  fmt.Sprintf("%s in-addr.arpa ip6.arpa", clusterDomain),
+						"configBlock": "pods insecure\nfallthrough in-addr.arpa ip6.arpa\nttl 30",
+					},
+					{"name": "prometheus", "parameters": "0.0.0.0:9153"},
+					{"name": "forward", "parameters": fmt.Sprintf(". %s", upstreamNameserver)},
+					{"name": "cache", "parameters": "30"},
+					{"name": "loop"},
+					{"name": "reload"},
+					{"name": "loadbalance"},
+				},
+			},
+		},
+	}
+
+	if serviceIP != "" {
+		values["service"] = map[string]any{
+			"clusterIP": serviceIP,
+		}
+	}
+
+	err = manager.Enable("dns", values)
+	if err != nil {
+		return fmt.Errorf("failed to enable dns component: %w", err)
+	}
+
+	client, err := k8s.NewClient()
+	if err != nil {
+		return fmt.Errorf("failed to create kubernetes client: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	dnsIP, err := k8s.GetServiceClusterIP(ctx, client, "ck-dns-coredns", "kube-system")
+	if err != nil {
+		return fmt.Errorf("failed to get dns service: %w", err)
+	}
+
```

Comment:
These are part of the cluster config. Can you add a TODO to store them with `database.SetClusterConfig`?

I'm okay with not delaying this PR any longer, we will return to it at a later time